front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts

```
@@ -74,58 +82,107 @@ const getTrainrunSectionsByTrainrunId = (netzgrafikDto: NetzgrafikDto, trainrunI
+      if (seenSectionIds.has(section.id)) {
+        throw new Error('Cycle detected in trainrun');
+      }
+      seenSectionIds.add(section.id);
+
+      if (section.sourceNodeId === nodeId) {
+        orderedSections.push(section);
+        nodeId = section.targetNodeId;
+        section = sectionsByConnectedTargetPortId.get(section.targetPortId);
+      } else if (section.targetNodeId === nodeId) {
+        // Swap source and target, so that the rest of the conversion logic
+        // can assume sections are always ordered from source to target
+        orderedSections.push({
+          ...section,
+          sourceNodeId: section.targetNodeId,
+          targetNodeId: section.sourceNodeId,
+          sourcePortId: section.targetPortId,
+          targetPortId: section.sourcePortId,
+          sourceDeparture: section.targetDeparture,
+          targetDeparture: section.sourceDeparture,
+          sourceArrival: section.targetArrival,
+          targetArrival: section.sourceArrival,
+        });
+        nodeId = section.sourceNodeId;
+        section = sectionsByConnectedSourcePortId.get(section.sourcePortId);
+      } else {
+        throw new Error('Section is disconnected from previous section');
+      }
     }
-    seenSectionIds.add(section.id);
+
+    orderedSectionPaths.push(orderedSections);
+  }
+
+  // We should've seen all of the train run's sections by now
+  if (seenSectionIds.size !== sections.length) {
+    throw new Error('Trainrun graph search failed to find all sections');
   }
 
-  // If we haven't seen all sections belonging to the trainrun, it's because
-  // it's made up of multiple separate parts
-  if (orderedSections.length !== sections.length) {
-    throw new Error('Trainrun is not continuous');
+  if (orderedSectionPaths.length === 0) {
+    throw new Error('Trainrun has no path');
   }
 
-  return orderedSections;
+  // TODO: some train runs are made up of multiple parts. We should create one
```

IMO we should raise an error, but I'm not a stakeholder :)

---

Yes agree with that

---

Right. I originally kept the error, and then removed it because the TODO will soon get resolved (in an upcoming patch) and that at least allows us to import all trainruns from RNE's large file. But with the hack commits linked above, we can workaround this easily for now - so let's re-introduce the error for now.